open_lm/model.py

```diff
@@ -157,17 +167,29 @@ def __init__(self, layer_id, args: Params):
         )
         self.attention.seq_len = args.seq_len
 
-        # initialize weights trunc_normal(1/sqrt(fan_in))
-        std = 1.0 / math.sqrt(args.dim)
-        torch.nn.init.trunc_normal_(
-            self.feed_forward.w12.weight, std=std, a=-3 * std, b=3 * std
-        )
-        # scale init by depth as in https://arxiv.org/abs/1908.11365 -- worked slightly better.
-        std = 1.0 / math.sqrt(hidden_dim)
-        std = std / math.sqrt(2 * (layer_id + 1))
-        torch.nn.init.trunc_normal_(
-            self.feed_forward.w3.weight, std=std, a=-3 * std, b=3 * std
-        )
+        if args.ffn_type == "swiglu":
+            # initialize weights trunc_normal(1/sqrt(fan_in))
+            std = 1.0 / math.sqrt(args.dim)
+            torch.nn.init.trunc_normal_(
+                self.feed_forward.w12.weight, std=std, a=-3 * std, b=3 * std
+            )
+            # scale init by depth as in https://arxiv.org/abs/1908.11365 -- worked slightly better.
+            std = 1.0 / math.sqrt(hidden_dim)
+            std = std / math.sqrt(2 * (layer_id + 1))
+            torch.nn.init.trunc_normal_(
+                self.feed_forward.w3.weight, std=std, a=-3 * std, b=3 * std
+            )
+        elif args.ffn_type == "gelu":
+            # Copied from mosaic-ml default init.
+            torch.nn.init.kaiming_normal_(self._ff_w1.weight, a=0.0, mode="fan_in", nonlinearity="relu")
```

Comment:
@mitchellnw moving from offline discussion - do you want this to be trunc_normal_?

Reply:
Yea I think we should change this initialization to be the same as we do do above in swiglu. It's already fairly similar but we should probably just stick to one for consistency.

In general I think we should move away from 1/block_idx extra variance scaling to 1/total_depth extra variance scaling for all output layers -- that's what's recommended by depth-mup. Or at least we should ablate on this. My guess is that it doesn't make a very large difference.

Reply:
Concretely in code I would recommend we leave the biases as is but then do the following instead of Kaiming normal:

```
            std = 1.0 / math.sqrt(args.dim)
            torch.nn.init.trunc_normal_(
                self._ff_w1.weight, std=std, a=-3 * std, b=3 * std
            )
            std = 1.0 / math.sqrt(hidden_dim)
            std = std / math.sqrt(2 * (layer_id + 1))
            torch.nn.init.trunc_normal_(
                self._ff_w2.weight, std=std, a=-3 * std, b=3 * std
            )
```
and the changes I'm curious in ablating over would be
A) changing `layer_id + 1` to `args.num_layers`.
B) removing the line `std = std / math.sqrt(2 * (layer_id + 1))`

Reply:
okay let me push this change before we kick off the 1b tests then

Reply:
ok done